yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -307,8 +307,14 @@ private[yarn] class YarnAllocator(
       nodes: Array[String],
       racks: Array[String]): ContainerRequest = {
     nodeLabelConstructor.map { constructor =>
+      val labelExp = if ((racks != null && (!racks.isEmpty))
+        || (nodes != null && (!nodes.isEmpty))) {
+        null
+      } else {
+        labelExpression.orNull
+      }
       constructor.newInstance(resource, nodes, racks, RM_REQUEST_PRIORITY, true: java.lang.Boolean,
-        labelExpression.orNull)
+        labelExp)
```

Comment:
From my understanding, currently in your implementation if `nodes` or `racks` is not empty, label expression will not be worked even it is explicitly set through configuration.

IMO I would choose to set `nodes` and `racks` to null if label expression is configured, otherwise user will be confused why explicitly setting lab expression is not worked.